### PR TITLE
Cortx_multiple_data_pod_shutdown_opt_3

### PR DESCRIPTION
**Pull request by ankita-93**
# Problem Statement
- Problem statement

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
-  [ ] New/Affected tests are executed on Latest Build
-  [ ] Attach test execution logs
-  [ ] Collection tested and no collection error introduced (`pytest --local True --collect-only`)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
-  [ ] If change in any common function, make sure to update all calls and execute all affected tests.

# Documentation
  Checklist for Author
-  [ ] Changes done to ReadMe / WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: tests/ha/multi_pod_failure/test_multi_data_server_pod_failure.py [hunk 1]
```diff
@@ -0,0 +1,613 @@
...
+        LOGGER.info("Step 4: Services of data and server pods are in offline state")
+
+        LOGGER.info("Step 5: Check services status on remaining pods %s", remain_pod_list)
+        resp = self.hlth_master_list[0].get_pod_svc_status(pod_list=remain_pod_list, fail=False,
+                                                           pod_name=running_pod)
+        LOGGER.debug("Response: %s", resp)
+        assert_utils.assert_true(resp[0], resp)
+        LOGGER.info("Step 5: Services status on remaining pod are in online state")
+
+        LOGGER.info("Step 6: Perform IOs with variable sizes objects.")
```
Comment (ApurwaMohite) [inline tests/ha/multi_pod_failure/test_multi_data_server_pod_failure.py:500]: We should try read-verify on already written data in step 1.

File: tests/ha/multi_pod_failure/test_multi_data_server_pod_failure.py [hunk 2]
```diff
@@ -0,0 +1,613 @@
...
+            assert_utils.assert_true(resp[0], resp)
+        LOGGER.info("Step 4: Services of data and server pods are in offline state")
+
+        LOGGER.info("Step 5: Check services status on remaining pods %s", remain_pod_list)
+        resp = self.hlth_master_list[0].get_pod_svc_status(pod_list=remain_pod_list, fail=False,
+                                                           pod_name=running_pod)
+        LOGGER.debug("Response: %s", resp)
+        assert_utils.assert_true(resp[0], resp)
+        LOGGER.info("Step 5: Services status on remaining pod are in online state")
+
```
Comment (ApurwaMohite) [inline tests/ha/multi_pod_failure/test_multi_data_server_pod_failure.py:600]: Same as above.

**Review by ApurwaMohite — APPROVED** [review]

File: tests/ha/multi_pod_failure/test_multi_data_pod_failure.py [hunk 3]
```diff
@@ -2051,105 +1722,6 @@ def test_degraded_deletes_till_kpods_fail(self):
...
-                                                    nclients=2, nsamples=2)
-        assert_utils.assert_true(resp[0], resp[1])
-        LOGGER.info("Step 6: Performed WRITE/READ/Verify/DELETEs with variable sizes objects.")
-
-        LOGGER.info("Completed: Test to Verify degraded IOs after multiple (max K) pods "
-                    "(data and server) failures with network of node hosting them going down.")
-
     @pytest.mark.ha
     @pytest.mark.lc
     @pytest.mark.skip(reason="Not supported in RGW yet")
```
Comment (rahul-hatwar) [inline tests/ha/multi_pod_failure/test_multi_data_pod_failure.py:1727]: remove this skip marker

File: tests/ha/multi_pod_failure/test_multi_data_pod_failure.py [hunk 4]
```diff
@@ -2454,56 +2001,28 @@ def test_copy_object_during_kpods_down(self):
...
-        pod_list = self.node_master_list[0].get_all_pods(pod_prefix=const.POD_NAME_PREFIX)
-        LOGGER.info("Get pod names to be deleted")
-        self.pod_name_list = self.system_random.sample(pod_list, self.kvalue)
-        for count, pod_name in enumerate(self.pod_name_list):
-            count += 1
+        LOGGER.info("Step 3: Shutdown random %s (K) data pods by deleting deployment and "
+                    "verify cluster & remaining pods status", self.kvalue)
+        resp = self.ha_obj.delete_kpod_with_shutdown_methods(
+            master_node_obj=self.node_master_list[0], health_obj=self.hlth_master_list[0],
+            down_method=const.RESTORE_DEPLOYMENT_K8S, kvalue=self.kvalue)
```
Comment (rahul-hatwar) [inline tests/ha/multi_pod_failure/test_multi_data_pod_failure.py:2008]: we have to add event=event

File: tests/ha/multi_pod_failure/test_multi_data_pod_failure.py [hunk 5]
```diff
@@ -2190,92 +1762,67 @@ def test_copy_object_kpods_fail(self):
...
-        resp = self.ha_obj.create_bucket_copy_obj(event, s3_test_obj=s3_test_obj,
-                                                  bucket_name=self.bucket_name,
-                                                  object_name=self.object_name,
-                                                  bkt_obj_dict=bkt_obj_dict,
-                                                  put_etag=put_etag,
-                                                  bkt_op=False)
-        assert_utils.assert_true(resp[0], f"Failed buckets are: {resp[1]}")
-        LOGGER.info("Step 7: Performed copy of %s from already created/uploaded %s to %s and "
-                    "verified copy object etags", self.object_name, self.bucket_name, bucket3)
+        if CMN_CFG["dtm0_disabled"]:
```
Comment (sonalk0209) [inline tests/ha/multi_pod_failure/test_multi_data_pod_failure.py:1796]: how about uploading new object to already created bucket1 and copy to bucket 2 once in degraded mode even if "if " is not satisfied?

File: tests/ha/multi_pod_failure/test_multi_data_pod_failure.py [hunk 6]
```diff
@@ -2533,49 +2052,48 @@ def test_copy_object_during_kpods_down(self):
...
-        LOGGER.info("Step 9: Download the uploaded %s on %s & verify etags.", objectnew, bucketnew)
-        resp = s3_test_obj.get_object(bucket=bucketnew, key=objectnew)
-        LOGGER.info("Get object response: %s", resp)
-        get_etag = resp[1]["ETag"]
-        assert_utils.assert_equal(put_etag, get_etag, "Failed in verification of Put & Get Etag "
-                                                      f"for object {objectnew} of bucket "
-                                                      f"{bucketnew}.")
-        LOGGER.info("Step 9: Downloaded the uploaded %s on %s & verified etags.",
-                    objectnew, bucketnew)
+        if CMN_CFG["dtm0_disabled"]:
```
Comment (sonalk0209) [inline tests/ha/multi_pod_failure/test_multi_data_pod_failure.py:2067]: same as above

File: tests/ha/multi_pod_failure/test_multi_data_server_pod_failure.py [hunk 7]
```diff
@@ -0,0 +1,635 @@
...
+                LOGGER.info("Cleanup: Get the network interface up for %s ip", node_ip)
+                node_iface = self.ip_dict.get(node_ip)[0]
+                worker_obj = self.ip_dict.get(node_ip)[1]
+                worker_obj.execute_cmd(cmd=cmd.IP_LINK_CMD.format(node_iface, "up"),
+                                       read_lines=True)
+                resp = sysutils.check_ping(host=node_ip)
+                assert_utils.assert_true(resp, "Interface is still not up.")
+        if os.path.exists(self.test_dir_path):
+            sysutils.remove_dirs(self.test_dir_path)
+        # TODO: As cluster restart is not supported until F22A, Need to redeploy cluster after
```
Comment (sonalk0209) [inline tests/ha/multi_pod_failure/test_multi_data_server_pod_failure.py:187]: we may need to chnage this message as:
Will need DTM support for pod restart and recovery so need to redeploy

File: tests/ha/multi_pod_failure/test_multi_data_server_pod_failure.py [hunk 8]
```diff
@@ -0,0 +1,635 @@
...
+
+        LOGGER.info("Done: Teardown completed.")
+
+    # pylint: disable=too-many-locals
+    # pylint: disable=too-many-statements
+    @pytest.mark.ha
+    @pytest.mark.lc
+    @pytest.mark.tags("TEST-35792")
+    @pytest.mark.skip(reason="Buckets cruds won't be supported with DTM0")
+    def test_server_data_kpods_fail_during_ios(self):
```
Comment (sonalk0209) [inline tests/ha/multi_pod_failure/test_multi_data_server_pod_failure.py:225]: do we not have any test for "after" scenario?

Comment (ankita-93) [reply tests/ha/multi_pod_failure/test_multi_data_server_pod_failure.py:224]: NO. This during is the only test where we are performing shutdown of both server and data pods.

**Review by rahul-hatwar — APPROVED** [review]
LGTM

File: tests/ha/multi_pod_failure/test_multi_data_pod_failure.py [hunk 9]
```diff
@@ -2190,92 +1760,78 @@ def test_copy_object_kpods_fail(self):
...
 
-        LOGGER.info("Step 8: Download the uploaded %s on %s & verify etags.", object3, bucket3)
-        resp = s3_test_obj.get_object(bucket=bucket3, key=object3)
-        LOGGER.info("Get object response: %s", resp)
-        get_etag = resp[1]["ETag"]
-        assert_utils.assert_equal(put_etag, get_etag, "Failed in verification of Put & Get Etag "
-                                                      f"for object {object3} of bucket {bucket3}.")
-        LOGGER.info("Step 8: Downloaded the uploaded %s on %s & verified etags.", object3, bucket3)
+        LOGGER.info("Step 5: Download the copied objects & verify etags.")
+        for bkt, obj in bkt_obj_dict2.items():
```
Comment (sonalk0209) [inline tests/ha/multi_pod_failure/test_multi_data_pod_failure.py:1822]: don't we need to check for bkt_obj_dict only?

**Review by sonalk0209 — APPROVED** [review]

**Review by rahul-hatwar — APPROVED** [review]

**Comment by ankita-93** [comment]
@sonalk0209 @ApurwaMohite @rahul-hatwar 
Please find the collection logs ss here
image

**Pull request merged**